### ISSUE TITLE
Add dynamic attributes to the Plant List survey 

### ISSUE THREAD

**Issue by sacrevert** [NERC-CEH/irecord-app#177]
Depends on https://github.com/BiologicalRecordsCentre/iRecord/issues/772

**Comment by kazlauskis** [comment]
@sacrevert Is this still relevant? what are the fields that I should update?

**Comment by sacrevert** [comment]
@kazlauskis @DavidRoy A few things that may be relevant here for other issues (I have little idea what is currently going on in this area):

- Plant survey in the app accepts vascular plants and bryos (which is fine), but the custom attributes that appear at the record level relate only to vascular plants
- Normal occurrence records of vascular plants in the app don't have the custom attributes provided by the plant survey approach, but bryophyte records *do* have the custom attributes available
- For the bryophyte form on the website (https://www.brc.ac.uk/irecord/enter_bryophyte_records) the coded habitat attribute that currently appears as a custom attribute in the app at the occurrence level is now at the sample level, at the occurrence level, there is a new free text microhabitat attribute, it is this that should appear at the occurrence level in the app (although I don't know anything about how this is stored in the database)

**Comment by kazlauskis** [comment]
Thanks. All the current attributes in the app should be reviewed - this relates to the #211. I will leave it to you guys to specify which specific fields I should add or remove from the app. @kitenetter

**Comment by sacrevert** [comment]
All of the above changes would be desirable, although I can't comment on resourcing/prioritisation issues. For example, enabling dynamic attributes within the plant survey mode may not be worth the effort. Certainly having the vascular plant specific attributes that are only within the plant survey enabled for normal occurrences would be worthwhile though.

**Comment by sacrevert** [comment]
1. Not sure what is being classed as "standard biological record ones" here, but any record under the Plant List Survey receives options that were generally specified for vascular plants (the options under Status and Abundance). I think this historically applied to Stage too, but at some point we edited that so that bryo-appropriate terms were also included. If one records a bryophyte via the casual route, the attributes are mostly different. I assume this means that the same information is being stored in different places (e.g. the Fruit attribute for a bryophyte recorded as a casual record stored as different attribute to "Stage --> Sporophyte" if recorded under Plant List Survey; same for Sex v. Male/Female).
2. Status is missing; Abundance is subtly different (I think a different warehouse attribute is being used) (and I'm not sure whether the Sex attribute is currently compatible with BSBI standards @japonicus ?)
3. A casual bryophyte record in the app allows for `@smpAttr:208` (EUNIS category -- at least I assume it is this one if it matches the website), but it should also have `@smpAttr:30` to fully match the website `enter_bryophyte_records` and the standard BBS request to recorders (i.e. a free-text occurrence level microhabitat).
4. I agree that the easiest solution may be to have two list surveys, although could a list survey not have the ability to check the taxon group of an entered species and choose the attributes based on that? It's possible that plant recorders would want to record both vasculars and some common bryos on a grid-square survey, and I would have thought that the generic square-list survey would also benefit from responsive attributes (ofc, accept that this may be beyond current resourcing)

**Comment by kitenetter** [comment]
@sacrevert I think the difficulty here is that the plant and bryophyte survey structures used on the website are not compatible; plus there are limitations to how dynamic we can make it.

**Individual records: bryophytes**
As far as I can see, the app captures a very similar set of attributes to the website form, with these exceptions:
1. The app has a boolean attribute for "Fruit", whereas the website has a lookup list for "Stage"
2. The app has a single attribute for habitat (which stores the top-level of the EUNIS hierarchy), while the website has the full EUNIS hierarchy, plus a free text habitat at sample level, plus a free text microhabitat at occurrence level
3. The app does not have an attribute for altitude; the website does

AFAIK the above discrepancies can be resolved by changing the app and/or the website, we just need to decide what is needed.

**Individual records: vascular plants**
This is more difficult to resolve, because the website form is intended for semi-structured grid recording, and therefore has attributes that are not relevant to recording an individual record on the app. It seems to me that these website attributes are not appropriate for the app individual records:
- multi-valued list of recorders; time surveying; number of recorders

The termlists used in the app for abundance and stage are already customised to match the website (but see below re abundance).

So it seems to me that the only attributes on the web form that could be added to the app are:
1. Status (BSBI)
2. Altitude (if this is needed for one-off individual records)

AFAIK those two could be added to the app, for vascular plants only. If we do that, we could also remove the app attribute for sex, which is not used on the website form.

**Plant List survey**
The difficulty here is that we can't have two separate survey structures within one app list survey. So either we have to compromise on a single list survey that allows entry of a set of attributes that applies to both vascular plants and bryophytes, or we have to consider having two list surveys, one for vascular plants and one for bryophytes.

With the current Plant List survey, we have enabled the storing of a multi-value list of recorders, and we have enabled the use of the BSBI stage and abundance terms. We do not currently enable:
- Time surveying; Number of recorders; Status (BSBI); Altitude

AFAIK those could be added, but if so they would then appear as options for all plants in the survey, including bryophytes.

AFAIK we cannot offer a different set of attributes for vascular plants vs bryophytes within a single list-type survey.

So I think the main question here is do you want to add a new "Bryophyte List survey" that matches the website form structure?

Here are my notes on what is currently happening:
[Plant attributes on the app.pdf](https://github.com/user-attachments/files/24133411/Plant.attributes.on.the.app.pdf)

If there are questions it's probably best for us to discuss online.

**Separate query re abundance lookups**
In the app, when adding a single record for a vascular plant, the abundance lookup offers choices that seem to me to be helpful:

<img width="494" height="1022" alt="Image" src="https://github.com/user-attachments/assets/93c37eae-1c36-462c-ae0e-be7b3fe59d77" />

However, when adding abundance to the app Plant List survey, or to the website, the user is faced with a blank data entry box:

<img width="370" height="519" alt="Image" src="https://github.com/user-attachments/assets/4e5074f7-f7fe-4f14-9738-c27d38e0c29b" />

Should we use the single record lookup format for the survey list?

**Comment by kazlauskis** [comment]
@kitenetter Like you say, the issue here is that the Plant survey is currently tied to `enter-vascular-plants` form. 

It appears that the bryophyte web form supports sub-samples, so plant and bryophyte surveys both use a `sample->sub-sample->occurrence` hierarchy. 

If it's helpful, I should be able to add dynamic attributes to the Plant survey in the app, so bryophytes can have their specific attributes, BUT these attributes won't appear on the website because the Plant survey records are linked to the website's edit form. If we go down this route, then we could:

1) Update the plant survey website form to be dynamic and accept different species groups.
2) Alternatively, I could direct each Plant survey sub-sample to a different website form: for bryophytes, use `enter_bryophyte_records`, and for other plants, use `enter-vascular-plants`. This way, main sample would still link to the vascular plant form and I am uncertain about how this would affect the website experience if someone opens up the top survey sample and then wants to edit each individual occurrence, which some would be plants and some bryophytes.

**Comment by kazlauskis** [comment]
@sacrevert @kitenetter I am bringing the discussion back here from the email. 

So that we're clear, let's talk only about adding the dynamic attributes to the **Plant List** survey in this issue, so that it would allow using both plant and bryophyte attributes. For the **Species List** or the **Single** record surveys that also allow recording plants in the app I have now created this ticket #337.

For the Plant List survey in the app (`enter-vascular-plants` _survey_id=325_) we are trying to add attributes from (`enter_bryophyte_records` _survey_id=327_).

Answering Oli's points:


> 1. add dynamic attributes to the Plant survey in the app

Yes, I can add dynamic attributes to the Plant List survey in the app.

> 2. add an occurrence-level altitude attribute for bryophytes (presumably the same attribute currently in use for the `enter-vascular-plants` form on the website) [this will then be in conflict with the bryo form on the website that has a sampled level altitude attribute, but I have a note to change that]

Yes, I can add the altitude attribute. The current altitude attribute for bryophytes survey is at the top-sample level on the website, not at sub-sample or occurrence levels. Should I add the altitude attribute to the top-sample or sub-samples in the app? @sacrevert 

> 3. for vascular plants, add Status (BSBI) and occurrence level altitude (both already on web form). Also remove vascular plant app attribute for sex, which is not that useful and not on the web form.

@sacrevert, I believe you are referring to the Species List survey here. The Plant List survey already allows recording the Status and does not use the Sex field. Created a separate ticket for this #337.

> 4. Add sample level attributes “time surveying” and “number of recorders” to plant list survey. These are already on vascular plant web form; I need to add them to bryo web form.

I can add the "time surveying" and "number of recorders" attributes. We already have the Recorders attribute in the Plant List survey in the app. Should we remove the "Recorders" field and replace it with the "number of recorders" as a dropdown list? I believe capturing the recorder names and deriving the number from the list is more accurate than using ranges.

> 5. Switch app abundance attribute for vascular plants to that one in use on the web form (@occAttr:610)

As with point 3, the Plant List survey already uses the `occAttr:610` for abundance counts. Created a separate ticket for this #337.

@sacrevert Which other bryophyte attributes are we adding to the Plant List survey?

If we add the bryophyte attributes to the app, the bryophyte attributes won't show up when users will edit the records on the website using `enter-vascular-plants` web form. @kitenetter should someone implement the dynamic attributes for the website as well then?

**Comment by sacrevert** [comment]
1. great thanks
2. my suggestion was for this to be at the record (occurrence level). Whilst i see that the subsample might be more appropriate (same altitude for all records grouped by grid ref?), I don't see how this would be implemented in the app. Occurrence level altitudes might keep this simplest?
3. Yes correct, thank you.
4. Please use both. The reason we have the list of names and the overall number is that often recording groups are large, but recorders will only want to make one or two names associated with the records, amalgamating others under somethign generic like "BBS Field Trip". See grey supporting information on the website `enter-vascular-plants` form here e.g.:
<img width="437" height="286" alt="Image" src="https://github.com/user-attachments/assets/516e3d3d-55e4-4491-968e-d2288dd98a6e" />
 
5. Yes, correct, thank you.

The bryophyte attributes to be added to the Plant List Survey are all those available currently for bryophytes under the Single record survey (plus the altitude attribute discussed above under 2).

Regarding the final question directed at @kitenetter , can it be changed so that bryophyte records made with the app using Plant List Survey are edited via the `enter_bryophyte_records` form instead of `enter-vascular-plants`, even if they are mixed in the original sample with vascular plants, or does that introduce other complications? At face value it seems like the simplest option.

**Comment by sacrevert** [comment]
Personally I would be happy for @kazlauskis to use the first option of editing the sample.input_form field for bryophytes. I'm not very concerned about having bryos/vasculars in different samples here (unless someone thinks I should be?)

I don't quite see how the dynamic attributes option will be straightforward given that bryophytes have a whole suite of different attributes: I don't see how these can easily be displayed via the current vascular plant listing form (`enter-vascular-plants`) on the website.

Sent from Outlook for Android<https://aka.ms/AAb9ysg>

________________________________

**Comment by kitenetter** [comment]
@sacrevert said "I don't quite see how the dynamic attributes option will be straightforward"

Agreed, they aren't straightforward - I think we're getting muddled again between dynamic termlists (which are possible on the website forms) and dynamic attributes (which are not currently available on the website forms).

So it looks like the best option is for @kazlauskis to adjust the sample.input_form field for bryophytes.

**Comment by kazlauskis** [comment]
John and I discussed this separately and concluded that we can use dynamic website forms for this. The app will set a special value to bryophyte records. The `enter-vascular-plants` web form will then dynamically display bryophyte attributes. This allows users to submit a list of plants and bryophyte records and view or edit the entire list through a single web form. @sacrevert are you happy with this?

**Comment by sacrevert** [comment]
@kazlauskis @johnvanbreda not really, because my prior on this is that displaying all the bryophyte dynamic attributes within the "container" of the `enter-vascular-plants` form is going to result in display issues that will then require time-consuming debugging. I don't know whether @johnvanbreda can demonstrate that this is not the case on one of the our test sites, but my expectation is that this is going to cascade problems...

**Comment by sacrevert** [comment]
Ok, that seems more sensible. @johnvanbreda does this then need an issue on the iRecord website GitHub for someone to configure these additions to the "Plant records" tab on the enter-vascular-plants form?

**Comment by sacrevert** [comment]
@kazlauskis @kitenetter I think a while back I thought this development and some other issues were imminent, but I'm not sure what happened to delay them?

**Comment by kazlauskis** [comment]
@sacrevert @kitenetter I needed to switch to other projects. I have a bit of time left from the original project for support and development to cover this, so I will try to work on the bryophyte attributes this week.

**Comment by kazlauskis** [comment]
@sacrevert If I add the Species Survey bryophyte attributes to the Plant Survey, then some will be dropped and some new ones added. Should I keep Identified by, Status and Abundance fields? Anything else?

current:
<img width="339" height="379" alt="Image" src="https://github.com/user-attachments/assets/b51009cc-9034-4cf1-beed-9e25fbdbf5db" />
new:
<img width="344" height="541" alt="Image" src="https://github.com/user-attachments/assets/01a94eeb-8e55-4332-8e82-025ee0edd85b" />

**Comment by kazlauskis** [comment]
@sacrevert, can you confirm if the above attributes are what you need?

**Comment by sacrevert** [comment]
That is correct, we are moving to the "new" screenshot situation, although can we also keep the "Identified by" field please.

We also discussed above the addition of a occurrence-level altitude attribute for bryophytes (I think this should map to `attr577`, as that is what is used by the vascular plant form, unless we are supposed to be creating unique attributes every time we do something like this? @kitenetter @johnvanbreda ).

**Comment by kitenetter** [comment]
@sacrevert Jim Bacon used to advise me that it was advisable to create new attributes for new surveys if there was any chance that the surveys might change their view in future about what the attribute should contain, which could then lead to problems if different surveys took different views. I can't envisage a situation where BSBI and BBS would disagree about what should be in an altitude attribute, so it should be fine to use the same one, but equally I'm not aware of any downsides to creating a new one.

**Comment by kazlauskis** [comment]
I have finished this ticket now. For website related changes (if needed) we should create a separate ticket.

> For clarity - the proposal I've discussed with Karolis is not really dynamic, but instead to have 2 separate data entry grids, one for vascular plants as now, but a separate grid underneath the vascular plants specifically for bryophytes. The bryophyte grid can then be configured to have a completely separate set of custom attribute columns to the vascular plant one. Obviously we would need some instructions on the form to explain to put vascular plants in the first list and bryophytes in the second and we can alter the species search on each grid accordingly.

@johnvanbreda remind me if I need to set anything specific to the bryophyte sub-samples or occurrences for the web form to work?

**Comment by sacrevert** [comment]
@kazlauskis I believe that I have already created the corresponding website ticket here: https://github.com/BiologicalRecordsCentre/iRecord/issues/2083
